### ISSUE TITLE
WebkitGTK runs missing from wpt.fyi since Feb 23

### ISSUE THREAD

**Issue by DanielRyanSmith** [web-platform-tests/wpt#38723]
The last successful experimental run for WebKitGTK is [showing as Feb 22](https://wpt.fyi/runs?label=master&label=experimental&max-count=100&product=chrome&product=edge&product=firefox&product=safari&product=webkitgtk&view=subtest).

cc @clopez

**Comment by clopez** [comment]
Let's see:

**1. commit epochs/daily/2023-02-23_01H aka 17905990b9ea9784d161a6259b3f24776acd6d2f** 
  - This is the last run that worked.
  - The Job wpt-webkitgtk_minibrowser-nightly-reftest-4 at taskcluster took 98:52.05
(5932 seconds)
  - Link: https://github.com/web-platform-tests/wpt/runs/11536889629

**2. commit epochs/daily/2023-02-24_01H aka 1ba5cfed0836d1f7ea78b115ef6d9c7ed767b8b9**
  - Job wpt-webkitgtk_minibrowser-nightly-reftest-4 at taskcluster failed because it reached
the max time allowed for it to run (120 mins / 7200 seconds)
  - Link: https://github.com/web-platform-tests/wpt/runs/11565756671

**3. commit epochs/daily/2023-02-25_01H aka c9f81de4242294a0d694ecd5a63155acfe6bf49a**
  - Job wpt-webkitgtk_minibrowser-nightly-reftest-4 at taskcluster failed because it reached
the max time allowed for it to run (120 mins / 7200 seconds)
  - Link: https://github.com/web-platform-tests/wpt/runs/11591412274

**4. commit epochs/daily/2023-02-27_01H aka 4741956679232e6233ed81e86510073f231eb773**
  - Job wpt-webkitgtk_minibrowser-nightly-reftest-4 at taskcluster failed because it reached
the max time allowed for it to run (120 mins / 7200 seconds)
  - Link: https://github.com/web-platform-tests/wpt/runs/11612460736

**5. commit epochs/daily/2023-02-28_01H aka 5b72ee9b50ae6898c94fe3ab50f4c68d4572977c**
  - Job wpt-webkitgtk_minibrowser-nightly-reftest-4 at taskcluster failed because it reached
the max time allowed for it to run (120 mins / 7200 seconds)
  - Link https://github.com/web-platform-tests/wpt/runs/11642974132

**Comment by clopez** [comment]
Trying to check if the issue is caused because WebKitGTK somehow has become slower or is because that chunk now has more tests.

```
$ cat 2023-02-23_01H_live_backing.log|grep TEST_START|wc -l
5798
$ cat 2023-02-24_01H_live_backing.log|grep TEST_START|wc -l
5359
```
So at first sight it doesn't look like the run on 2023-02-24 executed more tests.
This may hint at an issue on WebKitGTK that for some reason has become slower running those tests or it can be also caused because the hardware resources assigned to the VM at takscluster have changed from one run to the other.

I will try to reproduce this locally to better see what is going on

**Comment by clopez** [comment]
Ok. I have left during the last 24 hours a machine dedicated (no other jobs in the meanwhile) to reproduce this doing two runs one after the other on a Ubuntu-20.04 container. 
- Run1: Checkout WPT git commit 1790599 and run the `wpt-webkitgtk_minibrowser-nightly-reftest-4` job with the same parameters than in takscluster using the WebKitGTK build from [260656@main](https://webkitgtk.org/built-products/x86_64/release/nightly/Ubuntu-20.04/MiniBrowser/MiniBrowser_gtk_260656@main.zip)
- Run2:  Checkout WPT git commit 1ba5cfe and run the `wpt-webkitgtk_minibrowser-nightly-reftest-4` job with the same parameters than in takscluster using the minibrowser from [gtk_260726@main](https://webkitgtk.org/built-products/x86_64/release/nightly/Ubuntu-20.04/MiniBrowser/MiniBrowser_gtk_260726@main.zip)

Basically trying to reproduce test cases 1. and 2. above and this are the times that I got:

- Run1: ```Average = 4162 sec. Times per run = [4820, 4845, 3726, 3737, 3710, 3745, 4556]```

- Run2: ```Average = 4210 sec. Times per run = [4827, 4837, 4599, 3767, 3985, 3761, 3697]```

I get that run2 is on average ~1% slower than run1.

So the conclusion that I get from this is that I can't reproduce any relevant slowness on the code worth investigating further.

Perhaps the issue is triggered because of different hardware configurations at taskcluster.
Don't know.

**Comment by clopez** [comment]
Possible solutions:

- Option 1: Enable parallel tests. We are using only one runner, if we pass to wpt the flag `--processes=N` we can speed up significantly the time this jobs take. But that maybe causes some issue with some test results? I also wonder why this has still not be enabled by default
- Option 2: Rise the `maxRunTime` timeout from 7200 (2h) to 10800 (3h) 
- Option 3: Rise the number of chunks for `wpt-reftest` from 5 to 6 so each chunk has less tests to handle.

Any other idea?

**Comment by clopez** [comment]
> * Option 1: Enable parallel tests. We are using only one runner, if we pass to wpt the flag `--processes=N` we can speed up significantly the time this jobs take. But that maybe causes some issue with some test results? I also wonder why this has still not be enabled by default

I can see that the servo tests are run with `--processes=12` since ea02cbf033bb3b066f4928f354cfc406519d15b3 (it seems to be the only browser running the WPT tests in parallel)

I can try to do a whole run of the WPT tests with WebKitGTK with one process and with several to see which differences on results/time I get. But maybe is faster/better that we just rise the maxRunTime or the number of chunks

Opinions?
//cc @jgraham

**Comment by clopez** [comment]
I have been doing some experiments running the tests in parallel and this provides a significant speed up even when the machine only has one core and it doesn't seem to cause unexpected results or extra timeouts.

I did the following experiments:

1. **Test 1:** Machine with 32 vcores (16 with HT) running the whole WPT test suite. I get this times and test results with WebKitGTK:

```
# Time to complete the whole wpt test run (machine with 32 vcores)
1 job   => 1072 mins
2 jobs  => 422 mins
4 jobs  => 189 mins
8 jobs  =>  95 mins
16 jobs => 57 mins
32 jobs => 30 mins

# Results for the whole test suite
1 job   => Ran 1849549 checks (1799312 subtests, 50237 tests) Expected results: 1764747 Unexpected results: 84730 test: 7147 (7 crash, 1133 error, 4873 fail, 11 precondition_failed, 1123 timeout)
2 jobs  => Ran 1849591 checks (1799354 subtests, 50237 tests) Expected results: 1764861 Unexpected results: 84658 test: 7130 (9 crash, 1133 error, 4852 fail, 11 precondition_failed, 1125 timeout)
4 jobs  => Ran 1849657 checks (1799420 subtests, 50237 tests) Expected results: 1764815 Unexpected results: 84770 test: 7126 (8 crash, 1136 error, 4851 fail, 11 precondition_failed, 1120 timeout)
8 jobs  => Ran 1849615 checks (1799378 subtests, 50237 tests) Expected results: 1764717 Unexpected results: 84826 test: 7125 (8 crash, 1137 error, 4860 fail, 11 precondition_failed, 1109 timeout)
16 jobs => Ran 1849537 checks (1799300 subtests, 50237 tests) Expected results: 1764615 Unexpected results: 84850 test: 7133 (9 crash, 1136 error, 4858 fail, 11 precondition_failed, 1119 timeout)
32 jobs => Ran 1849488 checks (1799251 subtests, 50237 tests) Expected results: 1763426 Unexpected results: 85990 test: 7159 (8 crash, 1132 error, 4856 fail, 11 precondition_failed, 1152 timeout)
```

We can conclude some things here:

 * Regarding the results: the results of the test suite (number of passed/failed/timeouts) doesn't change significatively, so we can conclude that using more parallel jobs will not have any negative effect on the results.

 * Regarding the time it takes: As expected using more parallel jobs make the test suite finish faster. This is expected as we are using more cores to run the workload.
   * However we can observe a non-linear increment of the time vs the number of jobs when the number of jobs is less than 4.


I suspected this non-linear increment could be caused because when running with less jobs the cost of waiting for the slow tests or the timeouts to happen is bigger since the runner has to wait for each one of them in serial instead of allowing the timeouts to happen in parallel.

To prove my point I ran a second test.

2. **Test 2:** : Limit the number of cores of the machine to 1 and then try to run with different parallel jobs
This time instead of running the whole test suite I just executed the `4/5 chunk of reftest` tests (parameter to wpt
`--this-chunk=4 --total-chunks=5 --test-type=reftest`) (to not have to wait so much for everything to finish) with several number
of jobs, but this time I restricted the whole run to only using one core of the machine (by prefixing the run with command `taskset -c 0`) so I could be sure that the improvement is related with using more jobs in parallel even when only having one core on the machine and not with just using more cores in parallel.

And this are the results I got

```
# Time to complete wpt test: --this-chunk=4 --total-chunks=5 --test-type=reftest
# All tests restricted to only using one core of the machine (commands prefixed with "taskset -c0")
# Times are in mins:secs
1  jobs => 46:48
2  jobs => 23:14
3  jobs => 14:16
4  jobs => 11:20
5  jobs => 10:20
6  jobs => 10:10
7  jobs => 8:59
8  jobs => 9:12
9  jobs => 8:42
10 jobs => 8:59
11 jobs => 8:23
12 jobs => 8:40
24 jobs => 8:45

# Results for the 4/5 chunk of reftest
1 jobs => Ran 5798 checks (5798 tests) Expected results: 4684 Unexpected results: 1114 test: 1114 (1077 fail, 37 timeout)
2 jobs => Ran 5798 checks (5798 tests) Expected results: 4676 Unexpected results: 1122 test: 1122 (1086 fail, 36 timeout)
3 jobs => Ran 5798 checks (5798 tests) Expected results: 4666 Unexpected results: 1132 test: 1132 (1096 fail, 36 timeout)
4 jobs => Ran 5798 checks (5798 tests) Expected results: 4672 Unexpected results: 1126 test: 1126 (1090 fail, 36 timeout)
5 jobs => Ran 5798 checks (5798 tests) Expected results: 4664 Unexpected results: 1134 test: 1134 (1098 fail, 36 timeout)
6 jobs => Ran 5798 checks (5798 tests) Expected results: 4665 Unexpected results: 1133 test: 1133 (1097 fail, 36 timeout)
7 jobs => Ran 5798 checks (5798 tests) Expected results: 4665 Unexpected results: 1133 test: 1133 (1097 fail, 36 timeout)
8 jobs => Ran 5798 checks (5798 tests) Expected results: 4666 Unexpected results: 1132 test: 1132 (1096 fail, 36 timeout)
9 jobs => Ran 5798 checks (5798 tests) Expected results: 4665 Unexpected results: 1133 test: 1133 (1097 fail, 36 timeout)
10 jobs => Ran 5798 checks (5798 tests) Expected results: 4663 Unexpected results: 1135 test: 1135 (1099 fail, 36 timeout)
11 jobs => Ran 5798 checks (5798 tests) Expected results: 4663 Unexpected results: 1135 test: 1135 (1099 fail, 36 timeout)
12 jobs => Ran 5798 checks (5798 tests) Expected results: 4660 Unexpected results: 1138 test: 1138 (1102 fail, 36 timeout)
24 jobs => Ran 5798 checks (5798 tests) Expected results: 4655 Unexpected results: 1143 test: 1143 (1107 fail, 36 timeout)
```

Conclusions:

* Regarding the results: again we don't observe any meaningful difference. Some important remark: even when only allowing one core to be used having 24 jobs in parallel didn't caused more timeouts.
So risking causing unexpected timeouts because of using more jobs than cores doesn't seem to be a problem that happens in reality.

* Regarding the time that it takes: As you can see, using more than 8 jobs doesn't give any gain in total time because this test was restricted to only using 1 core of the machine so 8 seems to be the point were the test suite is able to squeeze the full power
of the 1 core assigned. And from 4 to 8 jobs the gain is not linear neither much relevant (only 25% less time when doubling the number of jobs).
However from 1 to 4 there is a really significant difference. From 1 to 4 jobs the speedup is 4.1X even when we only have one core assigned (restricted via taskset). Is almost as good as if we had 4 cores assigned.


Here is a graph [plotted with python with the above data](http://sprunge.us/y9YW23) where you can see this:

![Figure_1_24](https://user-images.githubusercontent.com/221181/225458635-85ee2954-0087-4f56-8900-b996fadc0bb3.png)

Zooming in:


![Figure_1_8](https://user-images.githubusercontent.com/221181/225458690-73bc3ad9-a3ea-467b-8de6-dc9692cf4abc.png)


So my conclusion is that running the tests with 4 parallel jobs is the sweet spot to overcome the lag caused by running slow tests in serial and waiting for tests timing out.

So I'm going to upload a PR proposing to run the webkitgtk tests with 4 parallel jobs on the CI.

**Comment by clopez** [comment]
Just uploaded PR at https://github.com/web-platform-tests/wpt/pull/39021

**Comment by DanielRyanSmith** [comment]
Thank you @clopez for fixing and also the detailed write-up to help understand. 🙂

**Comment by clopez** [comment]
Seems that it worked as expected:

- Before: [run of nightly-reftest-4](https://github.com/web-platform-tests/wpt/runs/12209882285): **115 mins**
- After: [run of nightly-reftest-4](https://github.com/web-platform-tests/wpt/runs/12240832998): **13 mins**

And the results are pretty much the same (differences likely to flakies or small changes in WebKit/wpt: they are testing different versions): https://wpt.fyi/results/?diff&filter=ADC&run_id=5181521781063680&run_id=5082498290155520
